lib/seattleflu/db/cli/command/etl/presence_absence.py

```diff
@@ -0,0 +1,388 @@
+                            identifier = test_result["id"],
+                            sample_id  = sample.id,
+                            target_id  = target.id,
+                            present    = get_target_result(test_result["targetStatus"], 
+                                                           sample.identifier, target.identifier),
+                            well_details = test_result["wellResults"])
+
+                mark_processed(db, group.id)
+
+                LOG.info(f"Finished processing presence_absence group {group.id}")
+
+    except Exception as error:
+        processed_without_error = False
+
+        LOG.error(f"Aborting with error")
+        raise error from None
+
+    else:
+        processed_without_error = True
+
+    finally:
+        if action == "prompt":
+            ask_to_commit = \
+                "Commit all changes?" if processed_without_error else \
+                "Commit successfully processed enrollments up to this point?"
+
+            commit = click.confirm(ask_to_commit)
+        else:
+            commit = action == "commit"
+
+        if commit:
+            LOG.info(
+                "Committing all changes" if processed_without_error else \
+                "Committing successfully processed enrollments up to this point")
+            db.commit()
+
+        else:
+            LOG.info("Rolling back all changes; the database will not be modified")
+            db.rollback()
+
+def find_or_create_target(db: DatabaseSession, identifier: str, control_status: bool) -> Any:
+    """
+    Select presence_absence test target by *identifier*, or insert it if it doesn't exist.
+    """
+    LOG.debug(f"Looking up target «{identifier}»")
+
+    target = db.fetch_row("""
+        select target_id as id, identifier
+          from warehouse.target
+         where identifier = %s
+        """, (identifier,))
+
+    if target:
+        LOG.info(f"Found target {target.id} «{target.identifier}»")
+    else:
+        LOG.debug(f"Target «{identifier}» not found, adding")
+
+        data = {
+            "identifier": identifier,
+            "control_status": control_status
+        }
+
+        target = db.fetch_row("""
+            insert into warehouse.target (identifier, control_status)
+                values (%(identifier)s, %(control_status)s)
+            returning target_id as id, identifier
+            """, data)
+
+        LOG.info(f"Created target {target.id} «{target.identifier}»")
+
+    return target
+
+
+def target_control_status(control_status: str, target_identifier: str) -> bool:
+    """
+    Determine the control status of the target.
+    """
+    expected_values = ["NotControl", "PositiveControl"]
+    if not control_status or control_status not in expected_values: 
+        raise UnknownControlStatusError(f"Unknown control status «{control_status}» for target {target_identifier}")
+    return control_status == "PositiveControl"
```

Comment:
As I mentioned in person but worth repeating: Really great to see this sort of defensive programming with explicit expectations for values, esp. in a data processing routine!